crates/ruff_linter/src/rules/flake8_comprehensions/fixes.rs

```diff
@@ -800,7 +803,211 @@ pub(crate) fn fix_unnecessary_map(
+
+    let expr = match object_type {
+        ObjectType::Generator => Expr::Generator(ast::ExprGenerator {
+            elt: placeholder.into(),
+            generators,
+            parenthesized: true,
+            range: TextRange::default(),
+        }),
+
+        ObjectType::List => Expr::ListComp(ast::ExprListComp {
+            elt: placeholder.into(),
+            generators,
+            range: TextRange::default(),
+        }),
+
+        ObjectType::Set => Expr::SetComp(ast::ExprSetComp {
+            elt: placeholder.into(),
+            generators,
+            range: TextRange::default(),
+        }),
+
+        ObjectType::Dict => Expr::DictComp(ast::ExprDictComp {
+            key: placeholder.clone().into(),
+            value: placeholder.into(),
+            generators,
+            range: TextRange::default(),
+        }),
+    };
+
+    let comment_ranges = checker.comment_ranges();
+    let source = checker.source();
+    let locator = checker.locator();
+
+    let mut content = generator.expr(&expr);
+
+    let original_expr_in_source = |expr: ExprRef, parent: AnyNodeRef| {
+        let original_range =
+            parenthesized_range(expr, parent, comment_ranges, source).unwrap_or(expr.range());
+
+        locator.slice(original_range)
+    };
+
+    let replacements: Vec<String> = match object_type {
+        ObjectType::Dict => {
+            fn parenthesized_if_necessary(expr: &Expr, as_str: String) -> String {
+                if expr.is_named_expr() {
+                    format!("({as_str})")
+                } else {
+                    as_str
+                }
+            }
+
+            let (key, value) = match body {
+                Expr::Tuple(ast::ExprTuple { elts, .. })
+                | Expr::List(ast::ExprList { elts, .. }) => {
+                    let [key, value] = &elts[..] else {
+                        bail!("Body is not key-value pair");
+                    };
+
+                    (key, value)
+                }
+                _ => bail!("Body is not key-value pair"),
+            };
+
+            let original_key = original_expr_in_source(key.into(), body.into());
+            let original_value = original_expr_in_source(value.into(), body.into());
+
+            vec![
+                parenthesized_if_necessary(key, original_key.to_string()),
+                parenthesized_if_necessary(value, original_value.to_string()),
+            ]
+        }
+
+        _ => {
+            let original_body = original_expr_in_source(body.into(), lambda.into());
+            vec![original_body.to_string()]
+        }
+    };
+
+    content = replace_placeholders(&content, &replacements);
+
+    // If the expression is embedded in an f-string, surround it with spaces to avoid
+    // syntax errors.
+    if matches!(object_type, ObjectType::Set | ObjectType::Dict) {
+        if parent.is_some_and(Expr::is_f_string_expr) {
+            content = format!(" {content} ");
+        }
+    }
```

Comment:
Do we also need to check whether the grandparent expression is an f-string? And the great-grandparent? And the great-great-grandparent? (Etc.)

Iterating over `semantic.current_expressions()` and checking whether any of them are f-string expressions might be more robust, if so

https://github.com/astral-sh/ruff/blob/638186afbd571c1a618bbcf356dd0f50785370ca/crates/ruff_python_semantic/src/model.rs#L1273-L1280